src/charm.py

```diff
@@ -5,52 +5,108 @@
         """Initialize the instance.
 
         Args:
             args: passthrough to CharmBase.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.config_changed, self.config_service)
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+        except CharmConfigInvalidError as exc:
+            self._update_status(BlockedStatus(exc.msg))
+            return
+        self._webserver = GunicornWebserver(
+            charm_state=self._charm_state,
+            flask_container=self.unit.get_container(FLASK_CONTAINER_NAME),
+        )
+        self._flask_app = FlaskApp(charm_state=self._charm_state)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.ingress = IngressPerAppRequirer(
             self,
-            port=self._FLASK_APP_PORT,
+            port=FLASK_APP_PORT,
             # We're forced to use the app's service endpoint
             # as the ingress per app interface currently always routes to the leader.
             # https://github.com/canonical/traefik-k8s-operator/issues/159
             host=f"{self.app.name}-endpoints.{self.model.name}.svc.cluster.local",
             strip_prefix=True,
         )
 
-    def config_service(self, event: ConfigChangedEvent) -> None:
+    def _update_status(self, status: StatusBase) -> None:
+        """Update the application and unit status.
+
+        Args:
+            status: the desired application and unit status.
+        """
+        self.unit.status = status
+        if self.unit.is_leader():
+            self.app.status = status
+
+    def container_can_connect(self) -> bool:
+        """Check if the Flask pebble service is connectable.
+
+        Returns:
+            True if the Flask pebble service is connectable, False otherwise.
+        """
+        return self.unit.get_container(FLASK_CONTAINER_NAME).can_connect()
+
+    def container(self) -> Container:
+        """Get the flask application workload container controller.
+
+        Return:
+            The controller of the flask application workload container.
+
+        Raises:
+            RuntimeError: if the pebble service inside the container is not ready while the
+                ``require_connected`` is set to True.
+        """
+        if not self.container_can_connect():
+            raise RuntimeError("pebble inside flask-app container is not ready")
+
+        container = self.unit.get_container(FLASK_CONTAINER_NAME)
+        return container
+
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
         """Configure the flask pebble service layer.
 
         Args:
             event: the config-changed event that trigger this callback function.
         """
-        container = self.unit.get_container("flask-app")
-        if not container.can_connect():
+        if not self.container_can_connect():
             event.defer()
             return
+
+        container = self.container()
```

Comment:
Just a nitpick, it feels a bit weird that self.container() already checks if it is connectable but we do a pre-check before.

Reply:
I updated the code to let the `container` method raises a `PebbleNotReadyError` and use try-catch for pebble availability issue. Thanks!